### PR TITLE
GEODE-3937:Fix NPE when executing removeFromDisk

### PR DESCRIPTION
**Pull request by dineshpune2006**
Fix NPE when executing removeFromDisk

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ X] Is your initial contribution a single, squashed commit?

- [ X] Does `gradlew build` run cleanly?

- [ X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.geode.test.junit.categories.DistributedTest;
+import org.awaitility.Awaitility;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+/**
+ * DUnit for ParallelGatewaySenderGeodeBug3937RegressionTest overflow operations.
+ */
+@Category(DistributedTest.class)
+public class ParallelGatewaySenderGeodeBug3937RegressionTest extends WANTestBase {
```
Comment (kirklund) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: I recommend renaming the test more in line with the way these RegressionTests are named:
- CreateAndLocalDestroyInTXRegressionTest
- EntriesDoNotExpireDuringGIIRegressionTest
- RemotePRValuesAreNotDeserializedRegressionTest
- ValuesAreLazilyDeserializedRegressionTest

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    long numMemVm4 = (Long) vm4.invoke(() -> WANTestBase.getNumberOfEntriesInVM("ln"));
+
+    long totalOverflown = numOvVm4 + numMemVm4;
+
+    assertEquals("Total number of entries  in VM4 is incorrect", (numEventPuts * 1),
+        (numOvVm4 + numMemVm4));
+
+    try {
+      vm4.invoke(() -> WANTestBase.clearGatewaySender("ln"));
+    } catch (Exception ee) {
```
Comment (kirklund) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: Remove all try-catch blocks from the test. JUnit's stack trace and error message is much better than using fail.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    vm4.invoke(
+        () -> WANTestBase.createPartitionedRegion(getTestMethodName(), "ln", 1, 100, isOffHeap()));
+
+    startSenderInVMs("ln", vm4);
+
+    vm4.invoke(() -> WANTestBase.pauseSender("ln"));
+
+
+    // give some time for the senders to pause
+    Wait.pause(1000);
```
Comment (kirklund) [inline geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: Replace this thread sleep with an Awaitility call. This sort of sleep/pause just creates Flaky tests.

Example:
```java
await().atMost(1, MINUTE).until(() -> vm4.invoke(() -> isSenderPaused("ln"));
```
such that you would add a new method called "isSenderPaused" which returns boolean if the sender is paused. Awaitility will poll that until it returns true or fail if it takes longer than 1 minute.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java [hunk 4]
```diff
@@ -1781,4 +1785,47 @@ ParallelGatewaySenderQueueMetaRegion newMetataRegion(InternalCache cache, final
...
+
+
+  public void clearQueueTestOnly() {
+
+    this.sender.pause();// it wil take internal read-write-lock
+    try {
+      for (PartitionedRegion prQ : this.userRegionNameToshadowPRMap.values()) {
+        clearPartitionedRegionTestOnly((PartitionedRegion) prQ);
+      }
+    } catch (Exception ee) {
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: Remove catch block that simply rethrows the caught exception.

File: geode-core/src/main/java/org/apache/geode/internal/cache/entries/DiskEntry.java [hunk 5]
```diff
@@ -1510,6 +1510,9 @@ public static void removeFromDisk(DiskEntry entry, LocalRegion region, boolean i
         throws RegionClearedException {
       DiskRegion dr = region.getDiskRegion();
       DiskId did = entry.getDiskId();
+      if (dr == null)
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/entries/DiskEntry.java]: Please use '{' brackets around all conditional bodies including one-liners:
```java
if (diskRegion == null) {
  return;
}
```

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java [hunk 6]
```diff
@@ -1781,4 +1785,47 @@ ParallelGatewaySenderQueueMetaRegion newMetataRegion(InternalCache cache, final
...
+    } finally {
+      if (this.sender.isPaused())
+        this.sender.resume();
+    }
+  }
+
+  // clear the partition region
+  private void clearPartitionedRegionTestOnly(PartitionedRegion partitionedRegion) {
+    LocalDataSet lds = (LocalDataSet) PartitionRegionHelper.getLocalPrimaryData(partitionedRegion);
+    Set<Integer> set = lds.getBucketSet(); // this returns bucket ids in the
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: Move multi-line end-of-line comment before the line. Spotless produces some undesirable results like this.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java [hunk 7]
```diff
@@ -1781,4 +1785,47 @@ ParallelGatewaySenderQueueMetaRegion newMetataRegion(InternalCache cache, final
...
+      if (bucket instanceof ProxyBucketRegion == false) {
+        if (bucket instanceof BucketRegion) {
+          BucketRegion bucketRegion = (BucketRegion) bucket;
+          Set keySet = bucketRegion.keySet();
+          for (Iterator iterator = keySet.iterator(); iterator.hasNext();) {
+            Object key = iterator.next();
+            try {
+              bucketRegion.remove(key);
+            } catch (Exception ee) {
+              ee.printStackTrace(System.out);
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: Remove all printStackTrace calls. If the exception is simply rethrown then remove the catch-block altogether.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java [hunk 8]
```diff
@@ -1781,4 +1785,47 @@ ParallelGatewaySenderQueueMetaRegion newMetataRegion(InternalCache cache, final
...
+    Set<Integer> set = lds.getBucketSet(); // this returns bucket ids in the
+                                           // function context
+
+    for (Integer bucketId : set) {
+      Bucket bucket = partitionedRegion.getRegionAdvisor().getBucket(bucketId);
+      if (bucket instanceof ProxyBucketRegion == false) {
+        if (bucket instanceof BucketRegion) {
+          BucketRegion bucketRegion = (BucketRegion) bucket;
+          Set keySet = bucketRegion.keySet();
+          for (Iterator iterator = keySet.iterator(); iterator.hasNext();) {
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: Nested for-loop, if-block, if-block, for-loop is too deep. Please combine the two if-blocks into one if-block. Consider breaking out if-block or inner for-loop to a new private method to improve readability.

File: geode-core/src/main/java/org/apache/geode/internal/cache/entries/DiskEntry.java (hunk 5 above)
Comment (dineshpune2006) [reply geode-core/src/main/java/org/apache/geode/internal/cache/entries/DiskEntry.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java (hunk 6 above)
Comment (dineshpune2006) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java (hunk 8 above)
Comment (dineshpune2006) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: done

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java (hunk 7 above)
Comment (dineshpune2006) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java]: done

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java (hunk 1 above)
Comment (dineshpune2006) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: Test File Name is changed now.

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java (hunk 3 above)
Comment (dineshpune2006) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: done

File: geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java (hunk 2 above)
Comment (dineshpune2006) [reply geode-wan/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderGeodeBug3937RegressionTest.java]: done

**Comment by dineshpune2006** [comment]
Hello Kirk, 
I made the changes as per review comments

**Comment by kirklund** [comment]
I added several additional reviewers for the asynceventqueue and wan code.

**Comment by dineshpune2006** [comment]
any comment?

**Comment by jhuynh1** [comment]
Hi

There are a few things going on in this PR… 

First it looks like you are attempting to add a lot of test code to clear a bucket region queue, this in turn ends up with an NPE thrown in DiskEntry.

However, Geode does not currently support clearing the bucket region queue and so this PR introduces some “test” product code.  The regression test then calls this “test” code to do the clear.  The NPE is then thrown.  The NPE itself looks like it would be a product issue, however it turns out that the remove you have attempted to call on the bucketRegion is not the correct remove when it comes to a bucketRegionQueue.  What should have been called is destroyKey from the bucket region queue.  Whenever clear is added to the product, it would have to handle this special case.

The fix to ignore the null in DiskEntry should not be added at this time.  If it were added with this test, the underlying entries would still probably reside in the queue.  The resume sender at the end of the test is why the queue ends up getting cleared.

I don’t think we should be adding this product code or the related “test” code.  It is hiliting a problem that does not currently exist in Geode.  The NPE is due to invalid test code.

here is are the entry that is removed when removing from a BucketRegionQueue:
[vm4] mapDestroy:EntryEventImpl[op=DESTROY;region=/__PR/_B__ln__PARALLEL__GATEWAY__SENDER__QUEUE_0;key=100;oldValue=null;newValue=null;callbackArg=null;originRemote=false;originMember=10.118.20.91(30103)<v1>:32773;id=EventIDid=18bytes;threadID=2;sequenceID=0]]

Compared to the entry the test code attempts to remove:
[vm4] mapDestroy:EntryEventImpl[op=DESTROY;region=/ln_PARALLEL_GATEWAY_SENDER_QUEUE;key=100;oldValue=null;newValue=null;callbackArg=null;originRemote=false;originMember=10.118.20.91(30117)<v1>:32773;id=EventIDid=18bytes;threadID=1;sequenceID=51]]

The BucketRegionQueue explicitly sets itself as the region to remove from, whereas the remove that was called in the test code does not do this.

**Comment by dineshpune2006** [comment]
i need to use 
BucketRegionQueue q=   this.getBucketRegionQueueByBucketId(partitionedRegion, bucketId);
					q.destroyKey(key);\
??

**Comment by jhuynh1** [comment]
You would need to do that if you want to remove a key from a bucketRegionQueue. Based on your test code it would be:

private void clearBucketRegionTestOnly(BucketRegion bucketRegion) {
    Set keySet = bucketRegion.keySet();
     for (Iterator iterator = keySet.iterator(); iterator.hasNext();) {
       Object key = iterator.next();
      ((BucketRegionQueue) bucketRegion).destroyKey(key);
     }
  }

However this pull request is about an NPE that does not exist in the product at this time, so I don't think this pull request should be committed

**Comment by dineshpune2006** [comment]
destroyKey it will remove the key & value. i am trying to clear the queue which meas. it will remove key & value holding inside a region which hold by async-queue. 

can i add clear method in ParallelGatewaySenderQueue.java and create another pull request??

**Comment by jhuynh1** [comment]
I think if you plan on adding a new api you would have to start a discuss topic on the dev list.  I would think you would want to add clear to all the gateway sender queues at that point but the change would have to be pretty thoroughly tested.  The change to call destroyKey was just a way to show you that the way that it was currently coded in the test code was not correct.  I am not sure if it is enough to just call that or if there is other book keeping that needs to be done.

**Comment by dineshpune2006** [comment]
thanks. i will start discussion on this  for adding the clear api.

**Comment by dineshpune2006** [comment]
will open another

**Pull request closed without merging**